src/core/server/legacy/legacy_service.ts

```diff
@@ -107,10 +109,34 @@ export class LegacyService implements CoreService {
       publishReplay(1)
     ) as ConnectableObservable<Config>;
 
-    this.configSubscription = update$.connect();
+    this.configSubscription = this.update$.connect();
+
+    const settings = await this.update$
+      .pipe(
+        first(),
+        map(config => getLegacyRawConfig(config))
+      )
+      .toPromise();
+
+    const { pluginSpecs, pluginExtendedConfig } = await findLegacyPluginSpecs(settings);
+
+    const uiExports = collectLegacyUiExports(pluginSpecs);
```

Comment:
yeah, I'm not so familiar with the underlying code, so not 100% confident here either. Whatever runs will also run when KbnServer initializes so I don't think it could introduce new behaviour somehow.

I tried to minimize the risk by only using these specs for migrations. Loading plugin specs https://github.com/elastic/kibana/blob/master/src/legacy/server/plugins/scan_mixin.js#L95 and parsing the uiExports https://github.com/elastic/kibana/blob/master/src/legacy/ui/ui_exports/ui_exports_mixin.js#L23 remains unchanged in KbnServer / Legacy.

This does mean we run `find_plugin_specs` twice. If there is code in a plugin's exported function that is stateful for some reason, this could change the behaviour of that plugin.

I searched for plugins that define variables outside the scope of the plugin function with `let` and couldn't find any. It's not a guarantee that all plugins are idempotent, but probably rules out the most common way.

Reply:
> This does mean we run find_plugin_specs twice. If there is code in a plugin's exported function that is stateful for some reason, this could change the behaviour of that plugin.

we can store found spec and reuse this value in Legacy server